pkg/destroy/aws/aws.go

```diff
@@ -77,1510 +63,1205 @@ func (o *ClusterUninstaller) validate() error {
-func getIAMClient(awsSession *session.Session) *iam.IAM {
-	return iam.New(awsSession)
+	return nil
 }
 
-// deleteNATGateways will attempt to delete all NAT Gateways that match the provided filters
-func deleteNATGateways(awsSession *session.Session, filters Filter, clusterName string, logger logrus.FieldLogger) (bool, error) {
-
-	logger.Debugf("Deleting NAT Gateways (%s)", filters)
-	defer logger.Debugf("Exiting deleting NAT Gateways (%s)", filters)
-
-	ec2Client := getEC2Client(awsSession)
-	describeNatGatewaysInput := ec2.DescribeNatGatewaysInput{}
-	describeNatGatewaysInput.Filter = createEC2Filters(filters)
-
-	// NAT Gateways take a while to really disappear so only find the ones not already being deleted
-	describeNatGatewaysInput.Filter = append(describeNatGatewaysInput.Filter, &ec2.Filter{
-		Name:   aws.String("state"),
-		Values: []*string{aws.String("available")},
+func deleteEC2InternetGateway(client *ec2.EC2, id string, logger logrus.FieldLogger) error {
+	response, err := client.DescribeInternetGateways(&ec2.DescribeInternetGatewaysInput{
+		InternetGatewayIds: []*string{aws.String(id)},
 	})
+	if err != nil {
+		return err
+	}
 
-	for {
-		results, err := ec2Client.DescribeNatGateways(&describeNatGatewaysInput)
-		if err != nil {
-			logger.Debugf("error listing NAT gateways: %v", err)
-			return false, nil
-		}
-
-		if len(results.NatGateways) == 0 {
-			break
-		}
-
-		for _, nat := range results.NatGateways {
-			logger.Debugf("deleting NAT Gateway: %v", *nat.NatGatewayId)
-			_, err := ec2Client.DeleteNatGateway(&ec2.DeleteNatGatewayInput{
-				NatGatewayId: nat.NatGatewayId,
+	for _, gateway := range response.InternetGateways {
+		for _, vpc := range gateway.Attachments {
+			_, err := client.DetachInternetGateway(&ec2.DetachInternetGatewayInput{
+				InternetGatewayId: gateway.InternetGatewayId,
+				VpcId:             vpc.VpcId,
 			})
-			if err != nil {
-				logger.Debugf("error deleting NAT gateway: %v", err)
-				continue
-			} else {
-				logger.WithField("id", *nat.NatGatewayId).Info("Deleted NAT Gateway")
+			if err == nil {
+				logger.WithField("vpc", *vpc.VpcId).Debug("Detached")
+			} else if err.(awserr.Error).Code() != "Gateway.NotAttached" {
+				return errors.Wrapf(err, "detaching from %s", *vpc.VpcId)
 			}
 		}
+	}
 
-		return false, nil
+	_, err = client.DeleteInternetGateway(&ec2.DeleteInternetGatewayInput{
+		InternetGatewayId: &id,
+	})
+	if err != nil {
+		return err
 	}
 
-	return true, nil
+	logger.Info("Deleted")
+	return nil
 }
 
-// deleteNetworkIface will attempt to delete a specific network interface
-func deleteNetworkIface(iface *string, ec2Client *ec2.EC2, logger logrus.FieldLogger) error {
-
-	result, err := ec2Client.DescribeNetworkInterfaces(&ec2.DescribeNetworkInterfacesInput{
-		NetworkInterfaceIds: []*string{iface},
+func deleteEC2NATGateway(client *ec2.EC2, id string, logger logrus.FieldLogger) error {
+	_, err := client.DeleteNatGateway(&ec2.DeleteNatGatewayInput{
```

Comment:
This drops the optimization https://github.com/openshift/installer/blob/d1272424396f1d42faf239dc1b976db905d0b338/pkg/destroy/aws/aws.go#L522

Reply:
> This drops the optimization...
>
>     // NAT Gateways take a while to really disappear so only find the ones not already being deleted

This commit tracks deleted ARNs, so it won't retry them like the old code.  However, there's no guard against running this delete against a NAT that someone else has deleted in parallel.  We could add that guard with something like our existing `err.(awserr.Error).Code() == "InvalidRouteTableID.NotFound"`, etc., but I haven't seen this hit an already-deleted NAT gateway in the wild, but it looks like it might be [`NatGatewayNotFound`][1].  Is that sufficient for adding it as an error check, or do you want to see if we can reproduce this against the real AWS API first?

[1]: https://github.com/openshift/installer/blob/v0.10.0/vendor/github.com/aws/aws-sdk-go/service/ec2/waiters.go#L837

Reply:
:+1: its fine for now I guess then.